polar-core/src/vm.rs

```diff
@@ -2579,12 +2638,29 @@ impl PolarVirtualMachine {
             // We're done; the rules are sorted.
             // Make alternatives for calling them.
 
-            self.polar_log_mute = false;
+            self.polar_trace_mute = false;
+            // as with the "QUERY RULE:" event print the first of these events
+            // to INFO and subsequent queries to TRACE (a superset of INFO)
+            let level = if self.queries.len() == 1 {
+                LogLevel::Info
+            } else {
+                LogLevel::Trace
+            };
             self.log_with(
+                level,
                 || {
                     let mut rule_strs = "APPLICABLE_RULES:".to_owned();
                     for rule in rules {
-                        rule_strs.push_str(&format!("\n  {}", self.rule_source(rule)));
+                        let context = self.kb.read().unwrap().get_rule_source(rule).map_or_else(
+                            || "".to_string(),
+                            |source| {
+                                let range = Range::from_span(&source.src, rule.span().unwrap());
+                                let context = Context { source, range };
+                                context.source_file_and_line()
+                            },
+                        );
+
+                        rule_strs.push_str(&format!("\n  {}{}", rule.head_as_string(), context));
```

Comment:
Don't necessarily need to figure this out for this PR, but it's slightly unfortunate that this is going to print out the rewritten ("longhand") head of a shorthand rule, like `has_permission(actor: Actor, "blorp", bleep: Bleep)`... although I guess it wouldn't be much clearer if we printed out the entire shorthand rule (`"blorp" if "blip";`). Maybe if we printed out the shorthand rule and underlined its head? Idk